tutorials/stats-sensor-space/76_new_cluster_test_api.py

```diff
@@ -0,0 +1,491 @@
+
+    # Shuffle the combined list
+    random.shuffle(combined_list)
+
+    # Separate the shuffled list back into participant IDs, conditions, and evoked data
+    shuffled_participant_ids, shuffled_conditions, shuffled_evoked_data = zip(
+        *combined_list
+    )
+
+    # Convert the tuples back to lists
+    shuffled_participant_ids = list(shuffled_participant_ids)
+    shuffled_conditions = list(shuffled_conditions)
+    shuffled_evoked_data = list(shuffled_evoked_data)
+
+    return shuffled_participant_ids, shuffled_conditions, shuffled_evoked_data
+
+
+def create_random_paired_evokeds_list():
+    """
+    Create a list of shuffled evoked data where each pair of target and non-target evoked data is shuffled together.
+    """
+    import random
+
+    _, evoked_data_a, evoked_data_b = prep_sample_data()
+
+    # Ensure evoked_data_a and evoked_data_b are of the same length
+    assert len(evoked_data_a) == len(
+        evoked_data_b
+    ), "evoked_data_a and evoked_data_b must have the same length"
+
+    # Create a list of participant indices
+    participant_indices = list(range(len(evoked_data_a)))
+
+    # Shuffle the list of participant indices
+    random.shuffle(participant_indices)
+
+    # Reorder evoked data according to the shuffled participant indices
+    shuffled_evoked_data_a = [evoked_data_a[i] for i in participant_indices]
+    shuffled_evoked_data_b = [evoked_data_b[i] for i in participant_indices]
+
+    # Combine the shuffled evoked data into a single list
+    shuffled_evoked_data = shuffled_evoked_data_a + shuffled_evoked_data_b
+
+    # Combine the original evoked data into a single list
+    original_evoked_data = evoked_data_a + evoked_data_b
+
+    return original_evoked_data, shuffled_evoked_data
+
+
+# shuffle order of pairs
+original_evoked_data, shuffled_evoked_data = create_random_paired_evokeds_list()
+# shouldn't change the results (p-value is different though?)
+
+shuffled_participant_ids, shuffled_conditions, shuffled_evoked_data = (
+    create_random_evokeds_id_condition_list()
+)
+
+
+def prepare_dataframe_for_cluster_function(
+    evokeds: list = None,
+    condition: list = None,
+    subject_index: list = None,
+):
+    """
+    Prepare a dataframe for the cluster test function.
+
+    Parameters
+    ----------
+    contrast : bool, optional
+        If True, a contrast is calculated. Default is False.
+    evokeds : list, optional
+        List of evoked objects. Default is None.
+    condition : list, optional
+        List of conditions for each evoked object. Default is None.
+    subject_index : list, optional
+        List of subject IDs. Default is None.
+
+    Returns
+    -------
+    df : DataFrame
+        The prepared DataFrame for the cluster test function.
+    """
+    # Initialize the DataFrame with evoked data
+    df = pd.DataFrame(
+        {
+            "evoked": evokeds,
+            "condition": condition if condition is not None else np.nan,
```

Comment:
consider using `pd.NA` instead of `np.nan` (np.nan will force the column type to be either `float` or `object`, whereas `pd.NA` allows conditions to be `int` or `str`)

Reply:
ah I didn't know the difference, thanks for pointing that out